### PR TITLE
Bisect FX node asserts on `ValidationException`.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107493
* #107989

This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

### CODE REVIEW

File: test/dynamo/test_exc.py [hunk 1]
```diff
@@ -195,47 +195,132 @@ def fn001(x):
...
+
+        @torch.compile
+        def fn(x, shape):
+            return x.split(shape)
+
+        self.assertExpectedInlineMunged(
+            ValidationException,
+            lambda: fn(torch.randn(20), (5, 10, 5)),
+            """\
+translation validation failed.
```
Comment (ezyang) [inline test/dynamo/test_exc.py:281]: How exactly am I supposed to interpret the message here?

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -263,6 +269,7 @@ def __init__(
                 allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
                 frame_id=frame_state["_id"],
                 co_fields=self.co_fields,
+                tracked_fakes=self.tracked_fakes,
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:272]: Mutable list sharing here is very delicate, call it out (and make sure it is tested!)

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
...
+            replacearg(index=1, key="args", fn=lambda args: tuple(convert_node(a) for a in args))
+        if self.is_evaluate_expr():
+            # ShapeEnv.evaluate_expr:
+            # "fx_node" parameter is an (optional) FX node that represents the evaluate expression.
+            # They must be replaced, since it will be part of a "call_function" FX node for
+            # torch._assert, which will be added to the FX graph of the new shape_env.
+            replacearg(index=3, key="fx_node", fn=convert_node)
+
+        # Actually call the method with the converted arguments.
+        return self.f(shape_env, *args, **kwargs)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2044]: This seems pretty delicate; in particular, it seems difficult to guarantee that the replay mechanism is going to keep working as people keep monkeying around with methods on ShapeEnv. First off, is this well tested? Second, one way to make it structurally harder to "mess it up" is to force all substantive method calls through this API, so that even regular, non bisecting ShapeEnv calls go through the logic here.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
...
+# Decorator for recording the given function as a replayable event.
+#
+# This decorator should be used at every function that mutates the state of
+# ShapeEnv in some way that affects the resulting issued guards (i.e. when
+# ShapeEnv.produce_guards is called).
+#
+# remember_fakes_length: flags whether we should add the length of the
+# ShapeEnv.tracked_fakes list to the event. This is used for calling
+# ShapeEnv.produce_guards at an arbitrary time (assuming the fakes list only
+# increases monotonically).
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2065]: If I add a new method which mutates the state of ShapeEnv, but I didn't read this code carefully enough, how will I know to apply this decorator? What test will tell me I did it wrong?

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -2108,15 +2223,36 @@ def __init__(
...
+                    "assume_static_by_default": assume_static_by_default,
+                    "specialize_zero_one": specialize_zero_one,
+                    "duck_shape": duck_shape,
+                    "frame_id": frame_id,
+                    "co_fields": co_fields,
+                    "should_record_events": False,
+                })
+            ]
+
+            self.tracked_fakes = tracked_fakes
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2247]: Note: this will almost certainly cause reference cycles, as tracked fakes has FakeTensor which has reference to SymInt which refer to ShapeEnv

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -2108,15 +2223,36 @@ def __init__(
...
+            self.events: List[ShapeEnvEvent] = [
+                ShapeEnvEvent(ShapeEnv, kwargs={
+                    "allow_scalar_outputs": allow_scalar_outputs,
+                    "allow_dynamic_output_shape_ops": allow_dynamic_output_shape_ops,
+                    "assume_static_by_default": assume_static_by_default,
+                    "specialize_zero_one": specialize_zero_one,
+                    "duck_shape": duck_shape,
+                    "frame_id": frame_id,
+                    "co_fields": co_fields,
+                    "should_record_events": False,
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2243]: If I add new arguments to the constructor, how do I know to add them here?

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -3400,6 +3580,10 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
                 eql, _ = self.create_fx_call_function(operator.eq, (fx_node, concrete_val))
                 node, fresh = self.create_fx_call_function(torch._assert, (eql,))
 
+            assert node is not None
+            if fresh:
+                node.meta["event"] = len(self.events) - 1
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3585]: what's this?

**Comment by ezyang** [comment]
I'm not convinced about the ongoing maintainability of the record-replay mechanism on top of ShapeEnv. It feels like it needs some preparatory refactors to justify itself. Try to convince me?

**Comment by ysiraichi** [comment]
Thank you for the review! You are right and your comments are on point. I will address all of them and try to make this PR more robust. I will ping you when I get it ready.

File: test/dynamo/test_exc.py (hunk 1 above)
Comment (ysiraichi) [reply test/dynamo/test_exc.py:281]: The problem that TV tries to solve is: _"Given **Assertions**, is there any **Model** that satisfies **Target Expressions**, but doesn't satisfy the source expressions?"_. The **Faiiled Source Expressions** represents an affirmative response, showing which source expressions that failed.

File: torch/fx/experimental/symbolic_shapes.py (hunk 7 above)
Comment (ysiraichi) [reply torch/fx/experimental/symbolic_shapes.py:3585]: This is so we know, when bisecting, what event corresponds to a given node. Say we want to replay until FX node `n`. This tells us until what event index we should replay to get to `n`.

File: test/dynamo/test_exc.py (hunk 1 above)
Comment (ysiraichi) [reply test/dynamo/test_exc.py:281]: Oh, I see. I thought the smaller problem instance would be a nice win already, e.g. easier to read and reason about. But that makes sense.

Comment (ysiraichi) [reply test/dynamo/test_exc.py:281]: Another thing I thought was that it would be nice to know under what program FX node we failed. But, is there any way to get this information at this point (inside `ShapeEnv`)?

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (ysiraichi) [reply torch/_dynamo/output_graph.py:272]: One alternative I can think of is: 

- New `ShapeEnv.tracked_fakes_length` field
- New `ShapeEnv.inc_tracked_fakes_length` function
    - Call it whenever `tracked_fakes` is modified

In order to make it less error-prone, what do you think of creating a class for automating that?

```python
class TrackedFakeList(list):
    def __init__(self, shape_env):
        self.shape_env = shape_env
        super().__init__()

    def append(self, obj):
        self.shape_env.inc_tracked_fakes_length()
        super().append(obj)
```

Comment (ezyang) [reply torch/_dynamo/output_graph.py:272]: I don't have a good alternative, what you have right now is OK, it's just delicate.

File: test/dynamo/test_exc.py (hunk 1 above)
Comment (ezyang) [reply test/dynamo/test_exc.py:281]: Yeah, the stack trace on the bad guard would be pretty useful.

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (ysiraichi) [reply torch/_dynamo/output_graph.py:272]: Would you rather have this circular reference? I kind of like this alternative idea...

Comment (ezyang) [reply torch/_dynamo/output_graph.py:272]: Alternate idea is fine too.

**Comment by ysiraichi** [comment]
@ezyang I'm thinking about splitting this PR into 2:
1. Introduce replay mechanism for `ShapeEnv`
2. Bisect FX node asserts on `ValidationException`

I think that would be easier to review.
What do you think?

**Comment by ezyang** [comment]
That's fine, do as many splits as is not annoying for you to do.

File: test/dynamo/test_exc.py (hunk 1 above)
Comment (ysiraichi) [reply test/dynamo/test_exc.py:281]: What do you think of this?

Comment (ezyang) [reply test/dynamo/test_exc.py:281]: Yes, this is better.

**Comment by ysiraichi** [comment]
@ezyang I finished refactoring this PR. I believe all tests should be passing now. Could you take a look at it again?
Here is a short summary of the changes:

- Record/Replay of `ShapeEnv` was moved to #107989
    - Created a method for comparing the state of a `ShapeEnv` instance
    - Added more comments about events recording and `ShapeEnv` state equality check
    - Added a dynamo config for checking if we recorded the events correctly in the beginning of `produce_guards`
    - Off by default, but on when running _test/dynamo/test_dynamic_shapes.py_
    - Main execution path goes through events `run` (instead of a `__call__` overload) method
- Keeping track of the currently executed FX node with a context manager: `set_current_node`
- Improved `BisectValidationException` message

File: torch/fx/experimental/recording.py [hunk 8]
```diff
@@ -163,6 +163,13 @@ def is_evaluate_expr(self) -> bool:
     def is_defer_runtime_assert(self) -> bool:
         return self.name == "defer_runtime_assert"
 
+    def getarg(self, *, index: int, key: str) -> Any:
```
Comment (ezyang) [inline torch/fx/experimental/recording.py:166]: What does this do? Can we get some docs?

Comment (ysiraichi) [reply torch/fx/experimental/recording.py:166]: This is so we can get an argument from an event. It gets an argument by its positional index or its name.

Comment (ysiraichi) [reply torch/fx/experimental/recording.py:166]: I guess we can get rid of it, though. It's not a good solution for getting arguments from events.

File: torch/fx/experimental/symbolic_shapes.py [hunk 9]
```diff
@@ -2412,6 +2412,11 @@ def remove_fx_node(self, node: Optional[torch.fx.Node]) -> None:
             self.name_to_node.pop(node.name)
             self.graph.erase_node(node)
 
+    def add_fx_node_metadata(self, node: torch.fx.Node) -> None:
+        from torch._dynamo.utils import get_current_node
+        node.meta["event"] = self.last_event_index()
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2417]: This is too generic. Meta dict is used by everyone. Give a more detailed string key

File: torch/fx/experimental/symbolic_shapes.py [hunk 10]
```diff
@@ -3203,7 +3208,9 @@ def issue_guard(guard: ShapeGuard) -> None:
 
         # First, issue all the non-trivial guards.
         for guard in self.guards:
-            if self._maybe_evaluate_static(guard.expr) is not None:
+            static_expr = self._maybe_evaluate_static(guard.expr)
+            if static_expr is not None:
+                self.log.debug("eval %s == %s [statically known]", guard.expr, static_expr)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3213]: Are you sure about this log? We don't log other `issue_guard` here. This seems like it could be confuisng.

File: torch/fx/experimental/validator.py [hunk 11]
```diff
@@ -559,18 +571,167 @@ def symbolstr(sym) -> str:
...
+        # Reconstruct shape_env until the event at event_number.
+        shape_env = replay_shape_env_events(events[:number + 1])
+        shape_env.graph.lint()
+        return check_shapeenv_fails(shape_env, events[number].tracked_fakes)
+
+    last_exception = check_shapeenv_fails(shape_env, shape_env.snapshot_tracked_fakes())
+
+    if not last_exception:
+        # We don't actually fail due to a produce_guards call.
+        # Stop and don't bisect.
```
Comment (ezyang) [inline torch/fx/experimental/validator.py:677]: maybe worth logging here?

File: torch/fx/experimental/validator.py [hunk 12]
```diff
@@ -559,18 +571,167 @@ def symbolstr(sym) -> str:
...
+        assert event.is_defer_runtime_assert(), f"unexpected event type: {event}"
+        failed_action = "adding runtime assert"
+
+    args = event.args
+    assert args is not None
+    assert len(args) >= 2, (
+        f"bisecting expects {event.name} to have at least 2 positional arguments. "
+        f"Got: {len(args)}"
+    )
+    sympy.Eq
```
Comment (ezyang) [inline torch/fx/experimental/validator.py:726]: ?

**Review by ezyang — APPROVED** [review]
